# 1. Мультимодальные конвейеры
### Задание 1.
Перейдите в Центр моделей https://huggingface.co/models?pipeline_tag=text-generation и щелкните соответствующий тег слева, чтобы отобразить только поддерживаемые модели для этой задачи.
Найдите 10 моделей, работающих с русским языком.
Поэкспериментируйте с ними. В качестве предложений проанализировать 7 предложений(3 - с меткой ПОЗИТИВ, 3 - НЕГАТИВ, 1 - НЕЙТРАЛ).  Сделайте вывод о качестве каждой модели в виде таблицы.
### Задание 2.
Поработайте с параметрами модели для данного класса задач из таблицы.

Сделайте вывод в виде таблицы по каждой модели и кадому параметру с точки зрения повышения качества анализа тональности текста.
Выберите наиболее значимые параметры, оказывающие влияние на качество.


In [ ]:
from transformers import pipeline

texts = [
    "Этот фильм просто великолепен!",
    "Я очень доволен результатом.",
    "Спасибо за отличную работу!",
    "Ужасное обслуживание, больше не приду.",
    "Товар оказался бракованным.",
    "Фильм скучный и затянутый.",
    "На улице идёт дождь."
]

expert_labels = ["POSITIVE", "POSITIVE", "POSITIVE", "NEGATIVE", "NEGATIVE", "NEGATIVE", "NEUTRAL"]

# Для сбора статистики
total = len(texts)

In [151]:
# Создаём глобальный словарь для хранения результатов
results = {}

def sentiment_analysis(model_id, model_name, texts=texts, expert_labels=expert_labels, results_dict=None):
    if results_dict is None:
        results_dict = results  # используем глобальный словарь по умолчанию

    classifier = pipeline("sentiment-analysis", model=model_name)
    
    print(f"\nМодель {model_id}: {model_name}")
    print("-" * 100)
    correct = 0
    total = len(texts)
    
    for text, expert in zip(texts, expert_labels):
        result = classifier(text)[0]
        model_label = result['label']
        score = result['score']
        match = 1 if model_label.upper() == expert else 0
        correct += match
        print(f"Текст: {text[:30]}... | Эксперт: {expert} | Модель: {model_label} ({score:.4f}) | Совпадение: {match}")
    
    percent = (correct / total) * 100
    # Кладём результат в словарь по ключу model_name (старое значение перезапишется)
    results_dict[model_name] = percent
    print(f"| Итоговый % совпадений модели с мнением эксперта {percent:.1f}")

In [154]:
model = "blanchefort/rubert-base-cased-sentiment"

sentiment_analysis(model_id=1, model_name=model)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 685.79it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: blanchefort/rubert-base-cased-sentiment
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 1: blanchefort/rubert-base-cased-sentiment
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9781) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7508) | Совпадение: 1
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: NEUTRAL (0.7171) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 100.0


In [155]:
model = "Tatyana/rubert-base-cased-sentiment-new"

sentiment_analysis(model_id=2, model_name=model)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 690.99it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: Tatyana/rubert-base-cased-sentiment-new
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 2: Tatyana/rubert-base-cased-sentiment-new
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9781) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7508) | Совпадение: 1
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: NEUTRAL (0.7171) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 100.0


In [156]:
model_name = "Kostya165/rubert_tiny2_russian_emotion_sentiment"
sentiment_analysis(model_id=3, model_name=model_name)

Loading weights: 100%|██████████| 57/57 [00:00<00:00, 643.79it/s, Materializing param=classifier.weight]                                     



Модель 3: Kostya165/rubert_tiny2_russian_emotion_sentiment
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: positive (0.8491) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: neutral (0.7429) | Совпадение: 0
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: positive (0.8342) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: neutral (0.4811) | Совпадение: 0
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: neutral (0.9374) | Совпадение: 0
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: neutral (0.8827) | Совпадение: 0
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: neutral (0.9360) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 42.9


In [157]:
model_name = "MonoHime/rubert-base-cased-sentiment-new"
sentiment_analysis(model_id=4, model_name=model_name)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 652.20it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: MonoHime/rubert-base-cased-sentiment-new
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 4: MonoHime/rubert-base-cased-sentiment-new
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9781) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: POSITIVE (0.9812) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7508) | Совпадение: 1
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: NEGATIVE (0.7516) | Совпадение: 1
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: NEUTRAL (0.7171) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 100.0


In [158]:
model_name = "seara/rubert-tiny2-russian-sentiment"
sentiment_analysis(model_id=5, model_name=model_name)

Loading weights: 100%|██████████| 57/57 [00:00<00:00, 704.56it/s, Materializing param=classifier.weight]                                     
BertForSequenceClassification LOAD REPORT from: seara/rubert-tiny2-russian-sentiment
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 5: seara/rubert-tiny2-russian-sentiment
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: positive (0.9895) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: positive (0.9905) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: positive (0.9944) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: negative (0.9545) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: neutral (0.5481) | Совпадение: 0
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: neutral (0.5787) | Совпадение: 0
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: neutral (0.8746) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 71.4


In [159]:
model_name = "r1char9/rubert-base-cased-russian-sentiment"
sentiment_analysis(model_id=6, model_name=model_name)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 735.38it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: r1char9/rubert-base-cased-russian-sentiment
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 6: r1char9/rubert-base-cased-russian-sentiment
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: positive (0.9940) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: positive (0.9547) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: positive (0.9903) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: negative (0.9408) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: negative (0.5390) | Совпадение: 1
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: neutral (0.6776) | Совпадение: 0
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: neutral (0.8940) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 85.7


In [160]:
model_name = "seara/rubert-base-cased-russian-emotion-detection-cedr"

expert = ["JOY", "JOY", "JOY", "SADNESS", "SADNESS", "SADNESS", "NO_EMOTION"]

sentiment_analysis(model_id=7, model_name=model_name, expert_labels=expert)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 691.62it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: seara/rubert-base-cased-russian-emotion-detection-cedr
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Модель 7: seara/rubert-base-cased-russian-emotion-detection-cedr
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: JOY | Модель: joy (0.9514) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: JOY | Модель: joy (0.9361) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: JOY | Модель: joy (0.9554) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: SADNESS | Модель: sadness (0.4638) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: SADNESS | Модель: no_emotion (0.9696) | Совпадение: 0
Текст: Фильм скучный и затянутый.... | Эксперт: SADNESS | Модель: sadness (0.8961) | Совпадение: 1
Текст: На улице идёт дождь.... | Эксперт: NO_EMOTION | Модель: no_emotion (0.9407) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 85.7


In [161]:
model_name = "SkyWater21/rubert-base-cased-ru-go-emotions-ekman"
expert = ["JOY", "JOY", "JOY", "SADNESS", "SADNESS", "SADNESS", "NO_EMOTION"]

sentiment_analysis(model_id=8, model_name=model_name, expert_labels=expert)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 705.78it/s, Materializing param=classifier.weight]                                      



Модель 8: SkyWater21/rubert-base-cased-ru-go-emotions-ekman
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: JOY | Модель: joy (0.9894) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: JOY | Модель: joy (0.9875) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: JOY | Модель: joy (0.9899) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: SADNESS | Модель: fear (0.6226) | Совпадение: 0
Текст: Товар оказался бракованным.... | Эксперт: SADNESS | Модель: neutral (0.9382) | Совпадение: 0
Текст: Фильм скучный и затянутый.... | Эксперт: SADNESS | Модель: anger (0.6106) | Совпадение: 0
Текст: На улице идёт дождь.... | Эксперт: NO_EMOTION | Модель: joy (0.4893) | Совпадение: 0
| Итоговый % совпадений модели с мнением эксперта 42.9


In [162]:
model_name = "kartashoffv/vashkontrol-sentiment-rubert"
sentiment_analysis(model_id=9, model_name=model_name)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 680.46it/s, Materializing param=classifier.weight]                                      



Модель 9: kartashoffv/vashkontrol-sentiment-rubert
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: Positive (0.9999) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: Positive (0.9999) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: Positive (0.9999) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: Neutral (0.7316) | Совпадение: 0
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: Negative (0.9603) | Совпадение: 1
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: Neutral (0.9786) | Совпадение: 0
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: Neutral (0.9947) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 71.4


In [163]:
model_name = "mxlcw/rubert-tiny2-russian-financial-sentiment"
sentiment_analysis(model_id=10, model_name=model_name)

Loading weights: 100%|██████████| 57/57 [00:00<00:00, 684.20it/s, Materializing param=classifier.weight]                                     



Модель 10: mxlcw/rubert-tiny2-russian-financial-sentiment
----------------------------------------------------------------------------------------------------
Текст: Этот фильм просто великолепен!... | Эксперт: POSITIVE | Модель: positive (0.9916) | Совпадение: 1
Текст: Я очень доволен результатом.... | Эксперт: POSITIVE | Модель: positive (0.9938) | Совпадение: 1
Текст: Спасибо за отличную работу!... | Эксперт: POSITIVE | Модель: positive (0.9844) | Совпадение: 1
Текст: Ужасное обслуживание, больше н... | Эксперт: NEGATIVE | Модель: negative (0.9627) | Совпадение: 1
Текст: Товар оказался бракованным.... | Эксперт: NEGATIVE | Модель: neutral (0.5470) | Совпадение: 0
Текст: Фильм скучный и затянутый.... | Эксперт: NEGATIVE | Модель: negative (0.8100) | Совпадение: 1
Текст: На улице идёт дождь.... | Эксперт: NEUTRAL | Модель: neutral (0.7550) | Совпадение: 1
| Итоговый % совпадений модели с мнением эксперта 85.7


In [183]:
print("\n" + "="*80)
print("ИТОГОВАЯ ТАБЛИЦА")
print("="*80)
print(f"{'Название модели':60} {'Итог, %':>7} {'Хорошо':^10}")
print("-" * (60+1+7+1+10))

for model, acc in results.items():
    good_flag = "Да" if acc > 80 else "Нет"
    print(f"{model[:60]:60} {acc:>6.1f}% {good_flag:^10}")


ИТОГОВАЯ ТАБЛИЦА
Название модели                                              Итог, %   Хорошо  
-------------------------------------------------------------------------------
blanchefort/rubert-base-cased-sentiment                       100.0%     Да    
Tatyana/rubert-base-cased-sentiment-new                       100.0%     Да    
Kostya165/rubert_tiny2_russian_emotion_sentiment               42.9%    Нет    
MonoHime/rubert-base-cased-sentiment-new                      100.0%     Да    
seara/rubert-tiny2-russian-sentiment                           71.4%    Нет    
r1char9/rubert-base-cased-russian-sentiment                    85.7%     Да    
seara/rubert-base-cased-russian-emotion-detection-cedr         85.7%     Да    
SkyWater21/rubert-base-cased-ru-go-emotions-ekman              42.9%    Нет    
kartashoffv/vashkontrol-sentiment-rubert                       71.4%    Нет    
mxlcw/rubert-tiny2-russian-financial-sentiment                 85.7%     Да    


### Задание 2

In [ ]:
model_name = "seara/rubert-base-cased-russian-emotion-detection-cedr"
text = "Этот фильм мне очень понравился, хотя начало было скучноватым, но потом всё раскрылось и актёры сыграли великолепно. Рекомендую к просмотру!"
classifier = pipeline("sentiment-analysis", model=model_name)

# Без параметров
res1 = classifier(text)[0]
print("Без параметров: ", res1)

# С top_k=3
res2 = classifier(text, top_k=3)
print("top_k=3:", res2)

# С truncation=True и max_length=30 (обрежем текст)
res3 = classifier(text, truncation=True, max_length=30)
print("truncation=True, max_length=30:", res3)

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 630.95it/s, Materializing param=classifier.weight]                                      
BertForSequenceClassification LOAD REPORT from: seara/rubert-base-cased-russian-emotion-detection-cedr
Key                          | Status     |  | 
-----------------------------+------------+--+-
bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Без параметров:  {'label': 'joy', 'score': 0.8585448265075684}
top_k=3: [{'label': 'joy', 'score': 0.8585448265075684}, {'label': 'sadness', 'score': 0.11807838827371597}, {'label': 'no_emotion', 'score': 0.033384405076503754}]
truncation=True, max_length=50: [{'label': 'joy', 'score': 0.7898435592651367}]


### вывод по заданию 2
Без параметров выдает что текст позитивный на 86%  
top_k=3 показывает что насколько текст позитивный, грустный и нейтральный  
после truncation обрезка текста до 30 показывает что текст радостный уже на 79%, значит обрезка текста сильно влияет на качество

# Классификации неразмеченных текстов
### Задание 3.
Перейдите в Центр моделей https://huggingface.co/models?pipeline_tag=text-generation и щелкните соответствующий тег слева, чтобы отобразить только поддерживаемые модели для этой задачи.
Найдите 10 моделей, работающих с русским языком.
Поэкспериментируйте с ними. В качестве предложений проанализировать 6 предложений (2 - одного класса, 2 - второго класса, 2 - третьего класса).   Сделайте вывод о качестве каждой модели в виде таблицы.
### Задание 4.
Поработайте с параметрами для модели данного класса задач.
